### PR TITLE
e2e: adding ibcwasm e2e upgrade test

### PR DESCRIPTION
**Pull request by damiannolan**
## Description

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: e2e/tests/wasm/upgrade_test.go [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+	err = chain.StartAllNodes(ctx)
+	s.Require().NoError(err, "error starting upgraded node(s)")
+
+	// we are reinitializing the clients because we need to update the hostGRPCAddress after
+	// the upgrade and subsequent restarting of nodes
+	s.InitGRPCClients(chain)
+
+	timeoutCtx, timeoutCtxCancel = context.WithTimeout(ctx, time.Minute*2)
+	defer timeoutCtxCancel()
```
Comment (colin-axner) [inline e2e/tests/wasm/upgrade_test.go:116]: I'm curious, there's now two defers in this function. Does this reset the context timeout?

**Comment by damiannolan** [comment]
Heads up - the job on this PR https://github.com/cosmos/ibc-go/actions/runs/7130649117/job/19417804360?pr=5333 running the chain upgrade test fails because the automated workflow doesn't provide the chain upgrade tag and plan name. 

When this PR is merged, I can rerun the E2E Upgrade workflow which should run all upgrade e2e tests. I've tested this locally with success:

```
=== NAME  TestIBCWasmUpgradeTestSuite/TestIBCWasmChainUpgrade
    testsuite.go:382: test passed, not uploading logs
    setup.go:210: Pruned 2 volumes, reclaiming approximately 14.8 MB
    setup.go:242: Pruned unused networks: [interchaintest-laulbrlj]
--- PASS: TestIBCWasmUpgradeTestSuite (801.24s)
    --- PASS: TestIBCWasmUpgradeTestSuite/TestIBCWasmChainUpgrade (801.24s)
        --- PASS: TestIBCWasmUpgradeTestSuite/TestIBCWasmChainUpgrade/create_and_exec_store_code_proposal (308.40s)
        --- PASS: TestIBCWasmUpgradeTestSuite/TestIBCWasmChainUpgrade/upgrade_chain (453.82s)
        --- PASS: TestIBCWasmUpgradeTestSuite/TestIBCWasmChainUpgrade/query_wasm_checksums (0.01s)
PASS
ok      github.com/cosmos/ibc-go/e2e/tests/wasm 802.086s
```

File: .github/workflows/e2e.yaml [hunk 2]
```diff
@@ -63,4 +63,4 @@ jobs:
       chain-b-tag: '${{ needs.determine-image-tag.outputs.simd-tag }}'
       chain-binary: 'simd'
       # on regular PRs we won't run upgrade tests.
-      test-exclusions: 'TestUpgradeTestSuite,TestGrandpaTestSuite'
+      test-exclusions: 'TestUpgradeTestSuite,TestGrandpaTestSuite,TestIBCWasmChainUpgrade'
```
Comment (damiannolan) [inline .github/workflows/e2e.yaml:66]: almost forgot about you :)

Comment (damiannolan) [reply .github/workflows/e2e.yaml:66]: nice catch!

**Pull request merged**